projects/main_ping_delays_pingsender.kp/knowledge.md

```diff
@@ -0,0 +1,468 @@
+
+for pre in PRES:
+    plot_cdf(delays_by_chan\
+             .filter(lambda d: d[0] == pre)\
+             .map(lambda d: d[1] / HOUR_IN_S if d[1] < MAX_DELAY_S else MAX_DELAY_S / HOUR_IN_S)\
+             .collect())
+    
+plt.legend(["No pingsender", "With pingsender"], loc="lower right")
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7fd034c91ad0>
+
+
+
+
+
+![png](images/output_24_1.png)
+
+
+The use of `pingsender` results in an improvement in the submission delay of the `shutdown` "main" ping. We receive almost 85% of the mentioned pings as soon as they are generated, instead of just ~30% within the first hour.
+
+We don't receive 100% of the pings sooner for builds having the `pingsender` enabled because the `pingsender` can fail submitting the ping (e.g. the system or Firefox uses a proxy, poor connection, ...) and, when this happen, no retrasmission is attempted; the ping will be sent on the next restart by Firefox.
+
+## How many duplicates come from the pingsender?
+Let's start by separating the pings coming from the `pingsender` from the ones coming from the normal Firefox flow since the `pingsender` started sending the `shutdown` pings. 
+
+
+```python
+post_pingsender_only = post_subset.filter(lambda p: p.get("meta/X-PingSender-Version") is not None)
+post_no_pingsender = post_subset.filter(lambda p: p.get("meta/X-PingSender-Version") is None)
+```
+
+```python
+num_from_pingsender = post_pingsender_only.count()
+num_no_pingsender = post_no_pingsender.count()
+total_post = post_subset.count()
+num_sent_by_both =\
+    post_pingsender_only.map(lambda p: p["id"]).intersection(post_no_pingsender.map(lambda p: p["id"])).count()
+```
+We want to understand how many pings were sent by the pingsender, correctly received from the server, and sent again next time Firefox starts.
+
+
+```python
+def pct(a, b):
+    return 100 * float(a) / b
+
+print("Duplicate pings percentage: {:.2f}%".format(pct(num_sent_by_both, total_post)))
+```
+    Duplicate pings percentage: 1.38%
+
+
+Do we get many more duplicates after landing the `shutdown pingsender`?
+
+
+```python
+count_deduped_pre = dedupe(pre_subset).count()
+count_pre = pre_subset.count()
+count_deduped_post = dedupe(post_subset).count()
+count_post = post_subset.count()
+
+print("Duplicates with shutdown pingsender:\nBefore:\t{:.2f}%\nAfter:\t{:.2f}%\n"\
+      .format(pct(count_pre - count_deduped_pre, count_pre),
+              pct(count_post - count_deduped_post, count_post)))
+```
+    Duplicates with shutdown pingsender:
+    Before:	0.50%
+    After:	1.61%
+    
+
+
+It looks like 1% of the pings sent by the `pingsender` are also being sent by Firefox next time it restarts. This is potentially due to `pingsender`:
+
+- being terminated after sending the ping but before successfully deleting the ping from the disk;
+- failing to remove the ping from the disk after sending it;
+- receiving an error code from the server even when the ping was successfully sent.
```

Comment:
I've updated the notebook to explain the difference, which is the one you correctly understood.

As for the pingsender saving extra stuff... we could discuss that, but I think it's unlikely for this to happen before it gets re-written in rust :)